File: src/chroma/code/typescript/querying.ts
// @snippet:imports
import { ChromaClient, type IncludeEnum } from 'chromadb';
import { DefaultEmbeddingFunction } from '@chroma-core/default-embed';

const client = new ChromaClient();
// @end

// @snippet:query-method

const embeddingFunction = new DefaultEmbeddingFunction();

const collection = await client.getOrCreateCollection({
  name: 'my_collection',
  embeddingFunction,
});

await collection.add({
  ids: ['doc1', 'doc2'],
  documents: [
    'Apples are really good red fruit',
    'Red cars tend to get more speeding tickets',
  ],
});

const results = await collection.query({
  queryTexts: ['I like red apples'],
});

const firstResult = results.documents[0];
// @end

// @snippet:query-with-options

const embeddingFunction2 = new DefaultEmbeddingFunction();

const collection2 = await client.getOrCreateCollection({
  name: 'my_collection',
  embeddingFunction: embeddingFunction2,
});

const results2 = await collection2.query({
  queryTexts: ['I like red apples'],
  nResults: 2,
  // specify what to include in the results
  include: ['distances', 'embeddings', 'metadatas'],
  // reduce the search space by only looking at these ids
  ids: ['id1', 'id2'],
  // filter results based on metadata
  where: { category: 'philosophy' },
  // filter results only to those matching document criteria
  whereDocument: { $contains: 'wikipedia' },
});

// return types

type QueryResult = {
  distances: (number | null)[][];
  documents: (string | null)[][];
  embeddings: (number[] | null)[][];
  ids: string[][];
  include: IncludeEnum[];
  metadatas: (Record<string, string | number | boolean> | null)[][];
};

type GetResult = {
  documents: (string | null)[];
  embeddings: number[][];
  ids: string[];
  include: IncludeEnum[];
  metadatas: (Record<string, string | number | boolean> | null)[];
};
// @end

// @snippet:metadata-filtering

await collection.query({
  queryTexts: ['first query', 'second query'],
  where: { page: 10 },
});

// In order to filter on metadata, you must supply a where filter dictionary to the query. The dictionary must have the following structure:
// {
//     metadata_field: {
//         <Operator>: <Value>
//     }
// }

// Using the $eq operator is equivalent to using the metadata field directly in your where filter.

const filter1 = {
  metadata_field: 'search_string',
};

// is equivalent to

const filter2 = {
  metadata_field: {
    $eq: 'search_string',
  },
};

const andExample = {
  $and: [
    {
      metadata_field1: {
        // <Operator>: <Value>
      },
    },
    {
      metadata_field2: {
        //<Operator>: <Value>
      },
    },
  ],
};

// @end

// @snippet:get-method

// Get by specific IDs
const docs = await collection.get({
  ids: ['doc1', 'doc2'],
});

// Get with pagination (default limit is 100)
const page = await collection.get({
  limit: 20,
  offset: 0,
});

// Get with metadata filter (no similarity ranking)
const filtered = await collection.get({
  where: { category: 'blog' },
  limit: 50,
});

// @end

// @snippet:where-document

// Find documents containing a specific string (case-sensitive)
const containsResults = await collection.query({
  queryTexts: ['search query'],
  whereDocument: { $contains: 'important keyword' },
});

// Exclude documents containing a string
const excludedResults = await collection.query({
  queryTexts: ['search query'],
  whereDocument: { $not_contains: 'deprecated' },
});

// Combine multiple document filters with $and
const combinedResults = await collection.query({
  queryTexts: ['search query'],
  whereDocument: {
    $and: [{ $contains: 'python' }, { $not_contains: 'legacy' }],
  },
});

// Combine whereDocument with metadata filtering
const fullFilterResults = await collection.query({
  queryTexts: ['search query'],
  nResults: 10,
  where: { status: 'published' },
  whereDocument: { $contains: 'tutorial' },
});

// @end

File: src/chroma/code/python/querying.py
# @snippet:imports
import os

import chromadb
from chromadb.api.types import Embeddings, ID, IDs, Document, Metadata, Include, EmbeddingFunction, Embeddable
from chromadb.utils.embedding_functions import OpenAIEmbeddingFunction
from typing import List, Optional, TypedDict, cast

client = chromadb.CloudClient(
    tenant=os.getenv("CHROMA_TENANT"),
    database=os.getenv("CHROMA_DATABASE"),
    api_key=os.getenv("CHROMA_API_KEY"),
)
# @end

# @snippet:query-method
embedding_function = OpenAIEmbeddingFunction(
	api_key="your-openai-api-key",
	model_name="text-embedding-3-large"
)

collection = client.get_or_create_collection(name="my_collection", embedding_function=cast(EmbeddingFunction[Embeddable], embedding_function))

collection.query(
    query_texts=["thus spake zarathustra", "the oracle speaks"],
)
# @end

# @snippet:query-with-options
embedding_function = OpenAIEmbeddingFunction(
	api_key="your-openai-api-key",
	model_name="text-embedding-3-large"
)

collection = client.get_or_create_collection(name="my_collection", embedding_function=cast(EmbeddingFunction[Embeddable], embedding_function))

collection.query(
    query_texts=["thus spake zarathustra", "the oracle speaks"],
	n_results=5,
	# specify what to include in the results
	include=["metadatas", "documents", "embeddings"],
	# reduce the search space by only looking at these ids
	ids=["id1", "id2"],
	# filter results only to those matching metadata criteria
	where={"category": "philosophy"},
	# filter results only to those matching document criteria
	where_document={"$contains": "wikipedia"},
)

# result shape

class QueryResult(TypedDict):
    ids: List[IDs]
    embeddings: Optional[List[Embeddings]]
    documents: Optional[List[List[Document]]]
    metadatas: Optional[List[List[Metadata]]]
    distances: Optional[List[List[float]]]
    included: Include

class GetResult(TypedDict):
    ids: List[ID]
    embeddings: Optional[Embeddings]
    documents: Optional[List[Document]]
    metadatas: Optional[List[Metadata]]
    included: Include
# @end

# @snippet:metadata-filtering

collection.query(
    query_texts=["first query", "second query"],
    where={"page": 10}
)

# In order to filter on metadata, you must supply a where filter dictionary to the query. The dictionary must have the following structure:
# {
#     "metadata_field": {
#         <Operator>: <Value>
#     }
# }


# Using the $eq operator is equivalent to using the metadata field directly in your where filter.
filter1 = {
    "metadata_field": "search_string"
}

# is equivalent to

filter2 = {
    "metadata_field": {
        "$eq": "search_string"
    }
}

and_example = {
    "$and": [
        {
            "metadata_field1": {
                # <Operator>: <Value>
            }
        },
        {
            "metadata_field2": {
                # <Operator>: <Value>
            }
        }
    ]
}

# @end

# @snippet:get-method

# Get by specific IDs
docs = collection.get(ids=["doc1", "doc2"])

# Get with pagination (default limit is 100)
page = collection.get(limit=20, offset=0)

# Get with metadata filter (no similarity ranking)
filtered = collection.get(
    where={"category": "blog"},
    limit=50
)

# @end

# @snippet:where-document

# Find documents containing a specific string (case-sensitive)
results = collection.query(
    query_texts=["search query"],
    where_document={"$contains": "important keyword"}
)

# Exclude documents containing a string
excluded = collection.query(
    query_texts=["search query"],
    where_document={"$not_contains": "deprecated"}
)

# Combine multiple document filters with $and
combined = collection.query(
    query_texts=["search query"],
    where_document={
        "$and": [
            {"$contains": "python"},
            {"$not_contains": "legacy"}
        ]
    }
)

# Combine where_document with metadata filtering
full_filter = collection.query(
    query_texts=["search query"],
    n_results=10,
    where={"status": "published"},
    where_document={"$contains": "tutorial"}
)

# @end